File: console/types/decision.ts
import type { GraphSubgraph, GraphSubgraphDto } from "@/types/graph";

export type DecisionTarget = {
  targetType: string;
  targetId: string;
};

export type DecisionEvidenceReference = {
  evidenceType: string;
  evidenceId: string;
  relationshipType: string | null;
  source: string | null;
  metadata: Record<string, unknown>;
};

export type DecisionPolicyReference = {
  policyId: string;
  policyVersion: string;
  policyName: string | null;
  metadata: Record<string, unknown>;
};

export type DecisionProvenance = {
  producerType: string;
  producerId: string;
  actorId: string | null;
  correlationId: string | null;
  requestId: string | null;
  createdAt: string;
};

export type DecisionSupersession = {
  supersedesDecisionId: string | null;
  supersededByDecisionId: string | null;
  supersessionReason: string | null;
};

export type GovernanceDecision = {
  decisionId: string;
  decisionType: string;
  status: string;
  target: DecisionTarget;
  reason: string;
  confidence: string;
  evidence: DecisionEvidenceReference[];
  policies: DecisionPolicyReference[];
  provenance: DecisionProvenance;
  supersession: DecisionSupersession;
  metadata: Record<string, unknown>;
  finalizedAt: string | null;
  archivedAt: string | null;
};

export type MissingEvidence = {
  evidenceType: string;
  reason: string;
  severity: string;
  metadata: Record<string, unknown>;
};

export type EvidenceNode = {
  entityType: string;
  entityId: string;
  label: string | null;
  lifecycle: string | null;
  attributes: Record<string, unknown>;
  metadata: Record<string, unknown>;
};

export type EvidenceEdge = {
  relationshipType: string;
  sourceType: string;
  sourceId: string;
  targetType: string;
  targetId: string;
  metadata: Record<string, unknown>;
};

export type DecisionEvidenceGraph = {
  targetType: string;
  targetId: string;
  nodes: EvidenceNode[];
  edges: EvidenceEdge[];
  missing: MissingEvidence[];
  metadata: Record<string, unknown>;
};

export type ReasoningEvidenceSummary = {
  targetType: string;
  targetId: string;
  metricScores: Record<string, number>;
  metricFailures: string[];
  thresholdBreaches: string[];
  driftSeverity: string | null;
  latestJobStatus: string | null;
  latestAuditStatus: string | null;
  evaluationResultIds: string[];
  metricIds: string[];
  driftAnalysisIds: string[];
  leaderboardIds: string[];
  jobIds: string[];
  mcpAuditIds: string[];
  policyIds: string[];
  missingEvidence: MissingEvidence[];
  metadata: Record<string, unknown>;
};

export type PolicyOutcome = {
  policyId: string;
  policyVersion: string;
  matchedRuleId: string | null;
  effect: string;
  reason: string;
  matched: boolean;
  metadata: Record<string, unknown>;
};

export type DecisionExplanation = {
  decisionId: string;
  summary: string;
  reasons: string[];
  evidenceReferences: DecisionEvidenceReference[];
  policyReferences: DecisionPolicyReference[];
  missingEvidence: MissingEvidence[];
  metadata: Record<string, unknown>;
};

export type DecisionAuditRecord = {
  auditId: string;
  decisionId: string;
  action: string;
  producerId: string;
  reason: string;
  actorId: string | null;
  correlationId: string | null;
  requestId: string | null;
  createdAt: string;
  metadata: Record<string, unknown>;
};

export type DecisionDetail = {
  decision: GovernanceDecision;
  evidenceSummary: ReasoningEvidenceSummary;
  policyOutcomes: PolicyOutcome[];
  auditRecords: DecisionAuditRecord[];
};

export type DecisionEvidence = {
  decisionId: string;
  evidenceReferences: DecisionEvidenceReference[];
  evidenceGraph: DecisionEvidenceGraph;
};

export type DecisionLineage = {
  decision: GovernanceDecision;
  subgraph: GraphSubgraph;
};

export type DecisionList = {
  decisions: GovernanceDecision[];
};

export type DecisionTargetDto = {
  target_type: string;
  target_id: string;
};

export type DecisionEvidenceReferenceDto = {
  evidence_type: string;
  evidence_id: string;
  relationship_type: string | null;
  source: string | null;
  metadata: Record<string, unknown>;
};

export type DecisionPolicyReferenceDto = {
  policy_id: string;
  policy_version: string;
  policy_name: string | null;
  metadata: Record<string, unknown>;
};

export type DecisionProvenanceDto = {
  producer_type: string;
  producer_id: string;
  actor_id: string | null;
  correlation_id: string | null;
  request_id: string | null;
  created_at: string;
};

export type DecisionSupersessionDto = {
  supersedes_decision_id: string | null;
  superseded_by_decision_id: string | null;
  supersession_reason: string | null;
};

export type GovernanceDecisionDto = {
  decision_id: string;
  decision_type: string;
  status: string;
  target: DecisionTargetDto;
  reason: string;
  confidence: string;
  evidence: DecisionEvidenceReferenceDto[];
  policies: DecisionPolicyReferenceDto[];
  provenance: DecisionProvenanceDto;
  supersession: DecisionSupersessionDto;
  metadata: Record<string, unknown>;
  finalized_at: string | null;
  archived_at: string | null;
};

export type MissingEvidenceDto = {
  evidence_type: string;
  reason: string;
  severity: string;
  metadata: Record<string, unknown>;
};

export type EvidenceNodeDto = {
  entity_type: string;
  entity_id: string;
  label: string | null;
  lifecycle: string | null;
  attributes: Record<string, unknown>;
  metadata: Record<string, unknown>;
};

export type EvidenceEdgeDto = {
  relationship_type: string;
  source_type: string;
  source_id: string;
  target_type: string;
  target_id: string;
  metadata: Record<string, unknown>;
};

export type DecisionEvidenceGraphDto = {
  target_type: string;
  target_id: string;
  nodes: EvidenceNodeDto[];
  edges: EvidenceEdgeDto[];
  missing: MissingEvidenceDto[];
  metadata: Record<string, unknown>;
};

export type ReasoningEvidenceSummaryDto = {
  target_type: string;
  target_id: string;
  metric_scores: Record<string, number>;
  metric_failures: string[];
  threshold_breaches: string[];
  drift_severity: string | null;
  latest_job_status: string | null;
  latest_audit_status: string | null;
  evaluation_result_ids: string[];
  metric_ids: string[];
  drift_analysis_ids: string[];
  leaderboard_ids: string[];
  job_ids: string[];
  mcp_audit_ids: string[];
  policy_ids: string[];
  missing_evidence: MissingEvidenceDto[];
  metadata: Record<string, unknown>;
};

export type PolicyOutcomeDto = {
  policy_id: string;
  policy_version: string;
  matched_rule_id: string | null;
  effect: string;
  reason: string;
  matched: boolean;
  metadata: Record<string, unknown>;
};

export type DecisionExplanationDto = {
  decision_id: string;
  summary: string;
  reasons: string[];
  evidence_references: DecisionEvidenceReferenceDto[];
  policy_references: DecisionPolicyReferenceDto[];
  missing_evidence: MissingEvidenceDto[];
  metadata: Record<string, unknown>;
};

export type DecisionAuditRecordDto = {
  audit_id: string;
  decision_id: string;
  action: string;
  producer_id: string;
  reason: string;
  actor_id: string | null;
  correlation_id: string | null;
  request_id: string | null;
  created_at: string;
  metadata: Record<string, unknown>;
};

export type DecisionDetailDto = {
  decision: GovernanceDecisionDto;
  evidence_summary: ReasoningEvidenceSummaryDto;
  policy_outcomes: PolicyOutcomeDto[];
  audit_records: DecisionAuditRecordDto[];
};

export type DecisionEvidenceDto = {
  decision_id: string;
  evidence_references: DecisionEvidenceReferenceDto[];
  evidence_graph: DecisionEvidenceGraphDto;
};

export type DecisionLineageDto = {
  decision: GovernanceDecisionDto;
  subgraph: GraphSubgraphDto;
};

export type DecisionListDto = {
  decisions: GovernanceDecisionDto[];
};

File: src/ai_governance/api/models/decision.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator

from ai_governance.api.models.ontology_graph import GraphSubgraphResponse


class DecisionEvaluateRequest(BaseModel):
    """
    REST request body for evaluating a governance decision.
    """

    target_type: str = Field(min_length=1)
    target_id: str = Field(min_length=1)
    decision_type: str = Field(min_length=1)
    policy_ids: list[str] = Field(default_factory=list)
    correlation_id: str | None = None
    request_id: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator(
        "target_type",
        "target_id",
        "decision_type",
        "correlation_id",
        "request_id",
    )
    @classmethod
    def string_must_not_be_blank(
        cls,
        value: str | None,
    ) -> str | None:
        if value is not None and not value.strip():
            raise ValueError("Decision request string fields must not be blank.")
        return value

    @field_validator("policy_ids")
    @classmethod
    def policy_ids_must_not_be_blank(
        cls,
        value: list[str],
    ) -> list[str]:
        if any(not item.strip() for item in value):
            raise ValueError("Decision policy_ids must not contain blanks.")
        return value


class DecisionTargetResponse(BaseModel):
    target_type: str
    target_id: str


class DecisionEvidenceReferenceResponse(BaseModel):
    evidence_type: str
    evidence_id: str
    relationship_type: str | None = None
    source: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class DecisionPolicyReferenceResponse(BaseModel):
    policy_id: str
    policy_version: str
    policy_name: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class DecisionProvenanceResponse(BaseModel):
    producer_type: str
    producer_id: str
    actor_id: str | None = None
    correlation_id: str | None = None
    request_id: str | None = None
    created_at: datetime


class DecisionSupersessionResponse(BaseModel):
    supersedes_decision_id: str | None = None
    superseded_by_decision_id: str | None = None
    supersession_reason: str | None = None


class DecisionAuditRecordResponse(BaseModel):
    audit_id: str
    decision_id: str
    action: str
    producer_id: str
    reason: str
    actor_id: str | None = None
    correlation_id: str | None = None
    request_id: str | None = None
    created_at: datetime
    metadata: dict[str, Any] = Field(default_factory=dict)


class DecisionResponse(BaseModel):
    decision_id: str
    decision_type: str
    status: str
    target: DecisionTargetResponse
    reason: str
    confidence: str
    evidence: list[DecisionEvidenceReferenceResponse]
    policies: list[DecisionPolicyReferenceResponse]
    provenance: DecisionProvenanceResponse
    supersession: DecisionSupersessionResponse
    metadata: dict[str, Any] = Field(default_factory=dict)
    finalized_at: datetime | None = None
    archived_at: datetime | None = None


class MissingEvidenceResponse(BaseModel):
    evidence_type: str
    reason: str
    severity: str
    metadata: dict[str, Any] = Field(default_factory=dict)


class EvidenceNodeResponse(BaseModel):
    entity_type: str
    entity_id: str
    label: str | None = None
    lifecycle: str | None = None
    attributes: dict[str, Any] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)


class EvidenceEdgeResponse(BaseModel):
    relationship_type: str
    source_type: str
    source_id: str
    target_type: str
    target_id: str
    metadata: dict[str, Any] = Field(default_factory=dict)


class DecisionEvidenceGraphResponse(BaseModel):
    target_type: str
    target_id: str
    nodes: list[EvidenceNodeResponse]
    edges: list[EvidenceEdgeResponse]
    missing: list[MissingEvidenceResponse]
    metadata: dict[str, Any] = Field(default_factory=dict)


class ReasoningEvidenceSummaryResponse(BaseModel):
    target_type: str
    target_id: str
    metric_scores: dict[str, float] = Field(default_factory=dict)
    metric_failures: list[str] = Field(default_factory=list)
    threshold_breaches: list[str] = Field(default_factory=list)
    drift_severity: str | None = None
    latest_job_status: str | None = None
    latest_audit_status: str | None = None
    evaluation_result_ids: list[str] = Field(default_factory=list)
    metric_ids: list[str] = Field(default_factory=list)
    drift_analysis_ids: list[str] = Field(default_factory=list)
    leaderboard_ids: list[str] = Field(default_factory=list)
    job_ids: list[str] = Field(default_factory=list)
    mcp_audit_ids: list[str] = Field(default_factory=list)
    policy_ids: list[str] = Field(default_factory=list)
    missing_evidence: list[MissingEvidenceResponse] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class PolicyOutcomeResponse(BaseModel):
    policy_id: str
    policy_version: str
    matched_rule_id: str | None = None
    effect: str
    reason: str
    matched: bool
    metadata: dict[str, Any] = Field(default_factory=dict)


class DecisionExplanationResponse(BaseModel):
    decision_id: str
    summary: str
    reasons: list[str]
    evidence_references: list[DecisionEvidenceReferenceResponse]
    policy_references: list[DecisionPolicyReferenceResponse]
    missing_evidence: list[MissingEvidenceResponse]
    metadata: dict[str, Any] = Field(default_factory=dict)


class ReasoningOutcomeResponse(BaseModel):
    decision: DecisionResponse
    explanation: DecisionExplanationResponse
    evidence_summary: ReasoningEvidenceSummaryResponse
    policy_outcomes: list[PolicyOutcomeResponse]


class DecisionDetailResponse(BaseModel):
    decision: DecisionResponse
    evidence_summary: ReasoningEvidenceSummaryResponse
    policy_outcomes: list[PolicyOutcomeResponse]
    audit_records: list[DecisionAuditRecordResponse]


class DecisionListResponse(BaseModel):
    decisions: list[DecisionResponse]


class DecisionEvidenceResponse(BaseModel):
    decision_id: str
    evidence_references: list[DecisionEvidenceReferenceResponse]
    evidence_graph: DecisionEvidenceGraphResponse


class DecisionLineageResponse(BaseModel):
    decision: DecisionResponse
    subgraph: GraphSubgraphResponse
